vfs/cache_local.go

```diff
@@ -0,0 +1,407 @@
+	var docs []*dirOrFile
+	sel := mango.Equal("folder_id", doc.ID())
+	req := &couchdb.FindRequest{Selector: sel, Limit: 10}
+	err = couchdb.FindDocs(c, FsDocType, req, &docs)
+	if err != nil {
+		return
+	}
+
+	for _, doc := range docs {
+		typ, dir, file := doc.refine()
+		switch typ {
+		case FileType:
+			lc.touchFile(file)
+			files = append(files, file)
+		case DirType:
+			lc.touchDir(dir)
+			dirs = append(dirs, dir)
+		}
+	}
+
+	return
+}
+
+// CreateFile is be used to persist a file document
+func (lc *LocalCache) CreateFile(c Context, doc *FileDoc) error {
+	err := couchdb.CreateDoc(c, doc)
+	if err != nil {
+		return err
+	}
+	lc.touchFile(doc)
+	return nil
+}
+
+// UpdateFile is used to update a persisted file document
+func (lc *LocalCache) UpdateFile(c Context, doc *FileDoc) error {
+	err := couchdb.UpdateDoc(c, doc)
+	if err != nil {
+		lc.rmFile(doc)
+		return err
+	}
+	lc.touchFile(doc)
+	return nil
+}
+
+// FileByID is used to fetch a file given its ID
+func (lc *LocalCache) FileByID(c Context, fileID string) (doc *FileDoc, err error) {
+	var ok bool
+	if doc, ok = lc.fileCachedByID(fileID); ok {
+		return
+	}
+
+	doc = &FileDoc{}
+	err = couchdb.GetDoc(c, FsDocType, fileID, doc)
+	if err != nil {
+		return nil, err
+	}
+
+	if doc.Type != FileType {
+		return nil, os.ErrNotExist
+	}
+
+	return doc, nil
+}
+
+// FileByPath is used to fetch a file given its path
+func (lc *LocalCache) FileByPath(c Context, name string) (doc *FileDoc, err error) {
+	dirpath := path.Dir(name)
+	parent, err := lc.DirByPath(c, dirpath)
+	if err != nil {
+		return
+	}
+
+	folderID, filename := parent.ID(), path.Base(name)
+
+	var ok bool
+	if doc, ok = lc.fileCachedByFolderID(folderID, filename); ok {
+		return
+	}
+
+	selector := mango.Map{
+		"folder_id": folderID,
+		"name":      path.Base(name),
+		"type":      FileType,
+	}
+
+	var docs []*FileDoc
+	req := &couchdb.FindRequest{
+		Selector: selector,
+		Limit:    1,
+	}
+	err = couchdb.FindDocs(c, FsDocType, req, &docs)
+	if err != nil {
+		return
+	}
+	if len(docs) == 0 {
+		err = os.ErrNotExist
+		return
+	}
+
+	doc = docs[0]
+	lc.touchFile(doc)
+	return
+}
+
+// DirOrFileByID is used to fetch a directory or file given its ID
+func (lc *LocalCache) DirOrFileByID(c Context, fileID string) (typ string, dirDoc *DirDoc, fileDoc *FileDoc, err error) {
+	var ok bool
+	if dirDoc, ok = lc.dirCachedByID(fileID); ok {
+		typ = DirType
+		return
+	}
+
+	if fileDoc, ok = lc.fileCachedByID(fileID); ok {
+		typ = FileType
+		return
+	}
+
+	dirOrFile := &dirOrFile{}
+	err = couchdb.GetDoc(c, FsDocType, fileID, dirOrFile)
+	if err != nil {
+		return
+	}
+
+	typ, dirDoc, fileDoc = dirOrFile.refine()
+	return
+}
+
+// Len returns the total number of elements currently cached
+func (lc *LocalCache) Len() int {
+	lc.mud.RLock()
+	lc.muf.RLock()
+	defer lc.mud.RUnlock()
+	defer lc.muf.RUnlock()
+	return lc.lrud.Len() + lc.lruf.Len()
+}
+
+func (lc *LocalCache) touchDir(doc *DirDoc) {
```

Comment:
I think we should rename this function. For me, `touch` is a write operation that changes the timestamp (like `touch` in the shell).